ferrocene/doc/user-manual/src/rustc/proxy-targets.rst

```diff
@@ -0,0 +1,273 @@
+      if even {
+         return 2;
+      } else {
+         return 1;
+      }
+   }
+
+First, we'll add support for the proxy target:
+
+.. code-block:: rust
+
+   #![cfg_attr(not(target_os = "linux"), no_std)]
+   #![no_main]
+
+   #[cfg(target_os = "linux")]
+   use std::ffi::c_int;
+   #[cfg(not(target_os = "linux"))]
+   use core::ffi::c_int;
+
+   #[unsafe(no_mangle)]
+   extern "C" fn return_a_number(even: bool) -> c_int {
+      if even {
+         return 2;
+      } else {
+         return 1;
+      }
+   }
+
+Next, we'll write a simple test suite using assertions that runs only on the proxy target:
+
+.. code-block:: rust
+
+   #[unsafe(no_mangle)]
+   fn main() {
+      assert_eq!(return_a_number(false), 1);
+      assert_eq!(return_a_number(true), 2);
+      println!("All tests passed successfully!");
+   }
+
+Build with the proxy target:
+
+.. code-block::
+   
+   rustc --target thumbv7em-ferrocenecoretest-eabihf src/thing.rs --out-dir artifacts -C instrument-coverage
+
+Then run it:
+
+.. code-block::
+
+   $ ./artifacts/thing 
+   All tests passed successfully!
+
+Try changing one of the assertions to be wrong then observe the failure:
+
+.. code-block::
+
+   $ rustc --target thumbv7em-ferrocenecoretest-eabihf src/thing.rs --out-dir artifacts
+   $ ./artifacts/thing 
+
+   thread '<unnamed>' panicked at src/thing.rs:21:5:
+   assertion `left == right` failed
+   left: 2
+   right: 3
+   note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
+   qemu: uncaught target signal 6 (Aborted) - core dumped
+   Aborted (core dumped)
+
+
+Code Coverage
+^^^^^^^^^^^^^
+
+We are going to write a simple library that can be both built as a library on a bare-metal target and
+as an executable on a proxy target.
+
+
+.. code-block:: rust
+   
+   // src/thing.rs
+
+   #![cfg_attr(not(target_os = "linux"), no_std)]
+   #![no_main]
+
+   #[cfg(target_os = "linux")]
+   use std::ffi::c_int;
+   #[cfg(not(target_os = "linux"))]
+   use core::ffi::c_int;
+
+   #[unsafe(no_mangle)]
+   extern "C" fn return_a_number(even: bool) -> c_int {
+      if even {
+         return 2;
+      } else {
+         return 1;
+      }
+   }
+
+   #[unsafe(no_mangle)]
+   fn main() {
+      assert_eq!(return_a_number(false), 1);
+      // assert_eq!(return_a_number(true), 2);
+   }
+
+
+   #[cfg(not(target_os = "linux"))]
+   #[panic_handler]
+   fn panic(_info: &core::panic::PanicInfo) -> ! {
+      loop {}
+   }
+
+We can build for a bare-metal target (:ref:`thumbv7em-none-eabihf`) by running:
+
+.. code-block::
+
+   rustc --target thumbv7em-none-eabihf src/thing.rs --out-dir artifacts
+
+We can build for the equivalent proxy target, with instrumentation, by running:
+
+.. code-block::
+   
+   rustc --target thumbv7em-ferrocenecoretest-eabihf src/thing.rs --out-dir artifacts -C instrument-coverage
+
+
+To create the ``profraw`` file:
+
+.. code-block::
+   
+   LLVM_PROFILE_FILE="thing.profraw" artifacts/thing
```

Comment:
You likely want to use at least `%p` (expands out to the process ID) and `%m` (expands out to the instrumented binary’s signature) in the profile filename to prevent overwriting profilings in case multiple tests are executed in one test run.

see https://clang.llvm.org/docs/SourceBasedCodeCoverage.html#running-the-instrumented-program